pkg/federate/kubefed/cluster.go

```diff
@@ -245,63 +260,31 @@ func newKubeFedClusterListerWatcher(kubeFedConfig *rest.Config) (cache.ListerWat
 	}, nil
 }
 
-func (c *clusterWatcher) onAdd(clusterID string, kubeConfig *rest.Config) {
-	c.enqueueEvent(func() {
-		c.handler.OnAdd(clusterID, kubeConfig)
-	}, clusterID, "Add", defaultEnqueueTimerInterval)
-}
-
-func (c *clusterWatcher) onUpdate(clusterID string, kubeConfig *rest.Config) {
-	c.enqueueEvent(func() {
-		c.handler.OnUpdate(clusterID, kubeConfig)
-	}, clusterID, "Update", defaultEnqueueTimerInterval)
-}
-
-func (c *clusterWatcher) onRemove(clusterID string) {
-	c.enqueueEvent(func() {
-		c.handler.OnRemove(clusterID)
-	}, clusterID, "Remove", defaultEnqueueTimerInterval)
-}
-
-func (c *clusterWatcher) enqueueEvent(event func(), clusterID string, eventType string, enqueueTimerInterval time.Duration) {
-	// First try to enqueue immediately to avoid the overhead of creating a Ticker for the common case.
-	select {
-	case c.eventQueue <- event:
-		return
-	default:
-	}
-
-	klog.V(2).Infof("Watcher for ClusterEventHandler \"%T\" eventQueue is full - starting timer loop", c.handler)
-
-	// Next we loop indefinitely trying to enqueue and preriodically log a warning.
-
-	startTime := time.Now()
-	ticker := time.NewTicker(enqueueTimerInterval)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case c.eventQueue <- event:
-			klog.Infof("Watcher for ClusterEventHandler \"%T\" successfully enqueued %s event for cluster \"%s\" after %fs",
-				c.handler, eventType, clusterID, time.Duration(time.Now().UnixNano()-startTime.UnixNano()).Seconds())
-			return
-		case <-ticker.C:
-			klog.Warningf("Watcher for ClusterEventHandler \"%T\" timed out after %fs trying to enqueue %s event for cluster \"%s\"",
-				c.handler, enqueueTimerInterval.Seconds(), eventType, clusterID)
-		case <-c.stopChan:
-			return
-		}
-	}
+func (c *clusterWatcher) enqueue(event clusterEvent) {
+	c.eventQueue.Add(event)
 }
 
 func (c *clusterWatcher) eventLoop() {
 	for {
-		select {
-		case event := <-c.eventQueue:
-			event()
-		case <-c.stopChan:
+		event, shutdown := c.eventQueue.Get()
+		if shutdown {
 			klog.V(2).Infof("ClusterEventHandler \"%T\" eventLoop stopped", c.handler)
 			return
 		}
+
+		event.(clusterEvent).run(c.handler)
```

Comment:
No action required: This interface (and generally the handler interface) won't allow for any error. IDK if it's what we'll really want in the future, but we can add it at a later point since it's only tangential to this patch..
Just popped into my mind so I'm writing it here to raise this design decision.

Reply:
yeah that thought crossed my mind as well.